### ISSUE TITLE
LUKS auto unlock via keyfile sometimes missing uuid

### ISSUE THREAD

**Issue by phillxnet** [rockstor/rockstor-core#2545]
Thanks to forum member Jorma_Tuomainen for highlighting this bug. Our Auto unlock option within Web-UI initiated LUKS full disk encryption currently fails to assign an associated UUID, resulting in the following file name:

> /root/keyfile-

Linking to the original pull request that added LUKS capability for reference and details of the prebug behaviour: #1716

See the following forum post for context: https://forum.rockstor.com/t/auto-unlock-with-keyfile/8779

**Comment by phillxnet** [comment]
Reproducer is to create a full disk LUKS container (disk format) via the instructions here:
https://rockstor.com/docs/howtos/luks.html#step-1-luks-format
Then set the Boot up Configuration as per the same howto **step 2:**
https://rockstor.com/docs/howtos/luks.html#step-2-boot-up-configuration

And we have, within text:

> Rockstor generated keyfile example: "/root/keyfile-fd168e30-5386-43b2-9f15-353b9ecff803". The characters after '-' are the uuid of the LUKS container ...

Yet our Web-UI indicates no "... uuid  of the LUKS container ..." in the 'Proposed Keyfile' name:

![no-luks-container-uuid-in-keyfile-name](https://user-images.githubusercontent.com/2521585/235350023-1c6ea9cc-6ff1-4141-b8e4-58a76ea6149d.png)

Where as we are expecting to see the proposed keyfile name ending in our LUKS container uuid. As per the graphics in our LUKS howto in the above Step 2.

Base OS "Uses openSUSE Tumbleweed Linux: 6.2.12-1-default"

**Comment by phillxnet** [comment]
This bug seems only to appear on first visiting this dialog, post having created the first luks device. A subsequent visit results in the expected behaviour, even with a freshly created luks container such as was done in the following:

![subsequent-visit-or-new-luks-disk-and-keyfile-name-ok](https://user-images.githubusercontent.com/2521585/235360548-acfd1da4-805b-4dd6-aa5b-9ddc7b0e73e4.png)

Working on reliable reproducer.

**Comment by phillxnet** [comment]
We are currently dealing only with the locked LUKS containers, we have yet to open one.

**Comment by phillxnet** [comment]
We may just be looking at a disk refresh here:

osi.get_whole_dev_uuid(dev_byid)
is a wrapper for "lsblk -n -o uuid dev-name-here"
with a disk freshly Web-UI LUKS formatted we have:
```
buildvm:~ # lsblk -n -o uuid /dev/disk/by-id/virtio-0002 
2f165ed9-5647-42a7-b46d-f3da41eda3d6
```
A consequent visit to LUKS disk dialog (cliicking on the locker icon by the patent disk "locked LUKS container") results in the expected use of this uuid in our keyfile option:

> Proposed Keyfile:  /root/keyfile-2f165ed9-5647-42a7-b46d-f3da41eda3d6
> WARNING: THE ABOVE KEY FILE DOES NOT EXIST

**Comment by phillxnet** [comment]
The root cause of this bug seems to be that sometimes we get the following disk (db model) role (db field) contents for the freshly created LUKS formatted drive. So it shows as a locked LUKS container without keyfile, but also has no uuid.

> {"LUKS": {"uuid": "", "keyfileExists": false, "unlocked": false}}

However this can be a tricky bug to 'catch' as it only happens sporadically. And a background disk refresh resolves this missing uuid, so there is a small window of time for it to emerge. I initially suspect we have a potential delay in "lsblk -n -o uuid dev" reporting the newly created (locked) LUKS container uuid:

https://github.com/rockstor/rockstor-core/blob/ac68e65e7eb9cb676052ce93bc4e5e8bf85e8e56/src/rockstor/storageadmin/views/disk.py#L792-L812

And for get_whole_dev_uuid(disk.name)

https://github.com/rockstor/rockstor-core/blob/ac68e65e7eb9cb676052ce93bc4e5e8bf85e8e56/src/rockstor/system/osi.py#L1952-L1981


Note the following in the above (fist extract) of
> @transaction.atomic
> def _luks_format(self, did, request, passphrase):

```
# TODO: Might be better to use cryptset luksUUID <dev-name> 
```

**Comment by phillxnet** [comment]
Referencing the TODO: we have the following output from the current test system:
```
buildvm:~ # cryptsetup luksUUID /dev/disk/by-id/virtio-0002
9b339276-bb75-4910-9791-8a29416004ca
buildvm:~ # lsblk -n -o uuid /dev/disk/by-id/virtio-0002
9b339276-bb75-4910-9791-8a29416004ca
```

**Comment by phillxnet** [comment]
To follow-up on the suspicion that lsblk may be behind, info wise, on the more area specific cryptsetup, re retrieving a uuid, I setup a sequential call to cryptsetup directly there-after that to lsblk for the same reason.

The following logging  shows the creation of a number of LUKS containers in order to see if cryptsetup can retrieve this uuid when lsblk fails:
```
[30/Apr/2023 20:07:02] DEBUG [system.osi:1986] get_whole_dev_uuid(virtio-0002) (/dev/disk/by-id/virtio-0002) returning dev_uuid=(1b229a57-6b03-42be-a7ee-59a44df93f76)
[30/Apr/2023 20:07:02] DEBUG [system.osi:1992] CRYPTSETUP returned out=['1b229a57-6b03-42be-a7ee-59a44df93f76', ''], err=[''], rc=0
...
[30/Apr/2023 20:07:36] DEBUG [system.osi:1986] get_whole_dev_uuid(virtio-0003) (/dev/disk/by-id/virtio-0003) returning dev_uuid=(22c4d49c-83db-4d23-b324-911f2bec93be)
[30/Apr/2023 20:07:36] DEBUG [system.osi:1992] CRYPTSETUP returned out=['22c4d49c-83db-4d23-b324-911f2bec93be', ''], err=[''], rc=0
...
 [30/Apr/2023 20:08:20] DEBUG [system.osi:1986] get_whole_dev_uuid(virtio-0004) (/dev/disk/by-id/virtio-0004) returning dev_uuid=(21d0b6bc-5a41-4864-aeac-57f2815481dd)
[30/Apr/2023 20:08:20] DEBUG [system.osi:1992] CRYPTSETUP returned out=['21d0b6bc-5a41-4864-aeac-57f2815481dd', ''], err=[''], rc=0
[30/Apr/2023 20:09:14] DEBUG [system.osi:1986] get_whole_dev_uuid(virtio-0005) (/dev/disk/by-id/virtio-0005) returning dev_uuid=()
[30/Apr/2023 20:09:14] DEBUG [system.osi:1992] CRYPTSETUP returned out=['cc65e822-b384-4b60-82f7-3a2115fe3d80', ''], err=[''], rc=0
```
The last instance shows cryptsetup was able to retrieve the uuid directly after lsblk failed to retrieve it. But of course this call to cryptsetup for uuid was after that of lsblk. But this may still be a more promissing approach, as per the TODO.

Also note that get_whole_dev_uuid(disk.name) is currently only used by this LUKS create related procedure.

**Comment by phillxnet** [comment]
Closing as:
Fixed by #2547
